### ISSUE TITLE
Custom directories in / (root) of jail

### ISSUE THREAD

**Issue by jooize** [iocage/iocage#271]
How can I create a directory like `/Volumes` in a jail (basejail)? The structure I used before and have in mind needs that.

I'd like to know any possible consequences, but mostly I just want it. :)

**Comment by skarekrow** [comment]
You can use a jailed dataset. Then you can create datasets underneath it for every mountpoint.

**Comment by jooize** [comment]
Not a ZFS dataset?

```
root@hipserv:/ # iocage create -b rootdir
  ERROR: invalid switch -b
root@hipserv:/ # iocage create tag=rootdir
  Successfully created: 65e2cff9-7ea1-11e5-ace3-f46d04d9905a (rootdir)
root@hipserv:/ # zfs create Hipstor/rootdir
root@hipserv:/ # iocage set jail_zfs=on rootdir
root@hipserv:/ # iocage set jail_zfs_dataset=Hipstor/rootdir rootdir
root@hipserv:/ # iocage start rootdir
* Starting 65e2cff9-7ea1-11e5-ace3-f46d04d9905a (rootdir)
  + Started (shared IP mode) OK
cannot mount '/Hipstor/rootdir': failed to create mountpoint
  + Starting services        OK
root@hipserv:/ # iocage console rootdir
root@65e2cff9-7ea1-11e5-ace3-f46d04d9905a:~ # zfs list
NAME              USED  AVAIL  REFER  MOUNTPOINT
Hipstor           322G   577G   104K  /Hipstor
Hipstor/rootdir    96K   577G    96K  /Hipstor/rootdir
root@65e2cff9-7ea1-11e5-ace3-f46d04d9905a:~ # ls /Hipstor
ls: /Hipstor: No such file or directory
root@65e2cff9-7ea1-11e5-ace3-f46d04d9905a:~ # mkdir /Volumes
mkdir: /Volumes: Read-only file system
root@65e2cff9-7ea1-11e5-ace3-f46d04d9905a:~ # zfs set mountpoint=/Volumes Hipstor/rootdir
root@65e2cff9-7ea1-11e5-ace3-f46d04d9905a:~ # ls /Volumes
ls: /Volumes: No such file or directory
```

**Comment by skarekrow** [comment]
Interesting. Thanks for the issue!

**Comment by skarekrow** [comment]
@jooize You should be able to update to the latest `develop` and use jailed datasets in the new basejails :)

**Comment by jooize** [comment]
I've landed in chaos. Not sure when what happened, unfortunately. I did attempt `upgrade` of some jails earlier, with unknown results. Let's begin with the issue relevant problem. Using iocage develop with latest commit c74c73909c36146ec01c14506654d6c24567c451.

Calling it `rootrid` instead of `rootdir` this time, which... is coincidental to another problem I discovered.

```
root@hipserv:~ # iocage create tag=rootrid
cannot rename 'Hipstor/iocage/base': dataset already exists
cannot rename 'Hipstor/iocage/releases': dataset already exists
  Successfully created: 3e2b59d5-84ce-11e5-8557-f46d04d9905a (rootrid)
root@hipserv:~ # zfs create Hipstor/rootdir
root@hipserv:~ # zfs destroy Hipstor/rootdir
root@hipserv:~ # zfs create Hipstor/rootrid
root@hipserv:~ # iocage set jail_zfs_dataset=Hipstor/rootrid rootrid
root@hipserv:~ # iocage set jail_zfs_mountpoint=/Volumes rootrid
cannot open 'Hipstor/iocage/jailsHipstor/rootrid': dataset does not exist
cannot open 'Hipstor/iocage/jailsHipstor/rootrid': dataset does not exist
cannot open 'Hipstor/iocage/jailsHipstor/rootrid': dataset does not exist
root@hipserv:~ # iocage start rootrid
* Starting 3e2b59d5-84ce-11e5-8557-f46d04d9905a (rootrid)
  + Started (shared IP mode) OK
  + Starting services        OK
root@hipserv:~ # iocage exec rootrid mount
Hipstor/iocage/jails/3e2b59d5-84ce-11e5-8557-f46d04d9905a/root on / (zfs, local, noatime, read-only, nfsv4acls)
root@hipserv:~ # iocage exec rootrid ls /Volumes
ls: /Volumes: No such file or directory
root@hipserv:~ # iocage console rootrid
<motd>
root@3e2b59d5-84ce-11e5-8557-f46d04d9905a:~ #
```

I have an empty entry in `iocage list`:
```
root@hipserv:~ # iocage list
JID  UUID                                  BOOT  STATE  TAG           TYPE
-    073b28f2-67d4-11e5-8f03-f46d04d9905a  on    down   dns           basejail
1    3e2b59d5-84ce-11e5-8557-f46d04d9905a  off   up     rootrid       basejail
<more jails>
-    -                                     -     down   -             -
```

Aaaand...
```
root@hipserv:~ # ls /iocage/tags/dns
_	fstab
```

I'm not sure whether `root` should always be there, and the important jail data is in `_`, right? Appears so. Either way I can live without the data. Have to begin using snapshots!

Here I find `854ff040-6dfc-11e5-9293-f46d04d9905a` which doesn't show up in `iocage list`, so perhaps that's the cause of the empty entry. That first snapshot is mine, of course.

```
root@hipserv:~ # zfs list -t snapshot
NAME                                                                             USED  AVAIL  REFER  MOUNTPOINT
Hipstor/iocage@2015-11-06                                                           0      -   220K  -
Hipstor/iocage/base/10.2-RELEASE/root@3e2b59d5-84ce-11e5-8557-f46d04d9905a          0      -   276M  -
Hipstor/iocage/base_old/10.2-RELEASE/root@073b28f2-67d4-11e5-8f03-f46d04d9905a      0      -   832M  -
Hipstor/iocage/base_old/10.2-RELEASE/root@d3c2e96a-67b3-11e5-953f-f46d04d9905a      0      -   832M  -
Hipstor/iocage/base_old/10.2-RELEASE/root@854ff040-6dfc-11e5-9293-f46d04d9905a      0      -   832M  -
Hipstor/iocage/base_old/10.2-RELEASE/root@0e1567a6-6ec7-11e5-9293-f46d04d9905a      0      -   832M  -
Hipstor/iocage/base_old/10.2-RELEASE/root@b6ce9001-7133-11e5-9ca8-f46d04d9905a      0      -   832M  -
Hipstor/iocage/base_old/10.2-RELEASE/root@c476e72c-7162-11e5-9739-f46d04d9905a      0      -   832M  -
Hipstor/iocage/base_old/10.2-RELEASE/root@e97a23f1-6de7-11e5-9293-f46d04d9905a      0      -   832M  -
Hipstor/iocage/base_old/10.2-RELEASE/root@b1bc7624-6876-11e5-9293-f46d04d9905a      0      -   832M  -
```

I'm using `hack88` with all jails except `rootrid`.

Can you get me out of this mess, or is it easier to start over? I have made copy–paste instructions for most of the jails, and it would be good to verify them anyway.

Hope you get something useful out, and ask away! :)

**Comment by skarekrow** [comment]
Any number of things could have happened :) Migration isn't solid yet from the older release, and it seems you have came from there, to here and back atleast 1 time. As for the usage, we are only supporting the `POOL/iocage/jails/UUID/data` currently in the develop branch. Which is likely what you are seeing in `iocage list`'s output. Please open a new ticket so I can make sure that different jailed dataset locations are supported. Quite a bit of things right now are under the assumption that your jailed dataset is /data.

I'm unsure of the importance of the data you have right now, but what I would do is this:

```
iocage clean -a
iocage fetch release=10.2-RELEASE
iocage create tag=something base=10.2-RELEASE  jail_zfs=on jail_zfs_mountpoint="/some/mount"
```

Then you can change that jailed datasets location by setting `jail_zfs_mountpoint` again.

**Comment by jooize** [comment]
I'll dig into this later. Thanks for your responsiveness and support! Am I correct in assuming you meant a new issue if `jail_zfs_mountpoint` fails after starting from scratch?

`iocage clean -a` is highly destructive to anyone reading looking for quick fixes.

> will destroy everything associated with iocage

**Comment by skarekrow** [comment]
No problem! Yes `iocage clean -a` is destructive, that's why I said I was unsure of the importance to your data ;) Anybody running the command should read the prompts before agreeing anyways.

No `jail_zfs_mountpoint` will work after starting from scratch. I meant opening an feature request to allow for jailed datasets to be named something other then `/data`. As right now that's hardcoded in and may or may not be trivial to change. But having an issue makes sure I don't forget to explore that option.

**Comment by jooize** [comment]
Noticed. Didn't check to see whether it's interactive. :p

Ahh, alright. I'll get to that too when I try this out!

Anyhow, as I only want to create a directory for mountpoints (e.g. `/Volumes/Disk{1,2}`), is jailing a dataset really the best way for that? You may have other things in mind, but it doesn't feel optimal for me. Occupies the `jail_zfs_dataset` slot I found useful too. What if someone wants more than one custom directory? You can't support *everything* of course, but I thought I'd mention it.

I'll probably get by fine with `jail_zfs_mountpoint=/Volumes` and `nullfs` mounts. I hope jail `fstab` mounts are applied after ZFS... x)

**Comment by skarekrow** [comment]
Thanks! They should mount after if you put `late` in the fstab. I can't remember if they are mounted before ZFS is. As for the jailed dataset, yes I do believe that is the best way to accomplish this. You can set the mountpoint for the root dataset inside the jail and then create new datasets under that dataset inside the jail and set the mountpoint for each directory anywhere you want. It will just require restarting the jail since it's r/o until it's rebooted and the jailed dataset is checked. 

I can't think of a better way to approach this as the basejails `/root` is not persistent between reboots of jails. Supporting additional unionfs directories for mounts sounds messy. I'd like to keep that directory to iocage itself, so we can rely on that not being tampered with. Or I guess we will just *hope* it isn't :D. 

Basically to recap, without using a jailed dataset, I cannot think of a clean way to let a user have their own mountpoints within a jail. This is a very good solution to that I believe.

For example: With the mountpoint set as `foo`
```
jail_zfs_mountpoint="/foo"
```
you can then start the jail and create more datasets with varying mountpoints.

Example:
```
iocage start JAIL
iocage console JAIL
zfs create -o mountpoint="/bar" POOL/iocage/jails/UUID/data
zfs create -o mountpoint="/baz" POOL/iocage/jails/UUID/data
exit
iocage reboot JAIL
PROFIT
```
Admittedly having to reboot the jail is less then ideal currently, and I plan on trying to remedy that.

**Comment by jooize** [comment]
Alright! Thanks for explaining. I've become occupied with other things, but I'll return to this eventually.

**Comment by skarekrow** [comment]
:+1:

**Comment by skarekrow** [comment]
@jooize you should no longer need to use a jailed dataset for this anymore. You can just do `mkdir /Volumes` inside the running jail with the `develop` from today. You can also still use the jailed dataset if you require it. But now you don't need to restart the jail to change the mountpoint of the jailed dataset inside of the running jail.

**Comment by jooize** [comment]
Excellent, thanks! How did you make it possible?

**Comment by skarekrow** [comment]
We moved from mounting each directory manually for the overlay to mounting the whole `_` over the jails `/root`. Giving the illusion of a full r/w jail!

**Comment by jooize** [comment]
Neat! :sunglasses:

**Comment by skarekrow** [comment]
Heh yep! Very neat.
